lib/Varien/Image/Adapter/Imagick.php

```diff
@@ -0,0 +1,252 @@
+    }
+
+    /**
+     * Destroy object image on shutdown
+     */
+    public function destruct()
+    {
+        if ($this->_imageHandler) {
+            $this->_imageHandler->destroy();
+        }
+    }
+
+    /**
+     * Opens image file.
+     *
+     * @param string $filename
+     * @throws ImagickException
+     * @throws Mage_Core_Exception
+     */
+    public function open($filename)
+    {
+        $this->_fileName = $filename;
+        $this->getMimeType();
+        $this->_getFileAttributes();
+
+        $this->_imageHandler = new Imagick();
+        $this->_imageHandler->readImage($filename);
+
+        $orientation = $this->_imageHandler->getImageProperty('exif:Orientation');
+        if (!empty($orientation)) {
+            switch ($orientation) {
+                case 1:
+                    // Do nothing
+                    break;
+                case 3:
+                    $this->_imageHandler->rotateImage('#000000', 180);
+                    break;
+
+                case 6:
+                    $this->_imageHandler->rotateImage('#000000', 90);
+                    break;
+
+                case 8:
+                    $this->_imageHandler->rotateImage('#000000', -90);
+                    break;
+                default:
+                    Mage::throwException('Unsupported EXIF orientation: ' . $orientation);
+            }
+        }
+        $this->refreshImageDimensions();
+    }
+
+    public function save($destination = null, $newName = null)
+    {
+        if ($destination && $newName) {
+            $fileName = $destination . "/" . $newName;
+        } elseif (isset($destination) && !isset($newName)) {
+            $info = pathinfo($destination);
+            $fileName = $destination;
+            $destination = $info['dirname'];
+        } elseif (!isset($destination) && isset($newName)) {
+            $fileName = $this->_fileSrcPath . "/" . $newName;
+        } else {
+            $fileName = $this->_fileSrcPath . $this->_fileSrcName;
+        }
+
+        $destinationDir = (isset($destination)) ? $destination : $this->_fileSrcPath;
+
+        if (!is_writable($destinationDir)) {
+            try {
+                $io = new Varien_Io_File();
+                $io->mkdir($destination);
+            } catch (Exception $e) {
+                throw new Exception("Unable to write file into directory '{$destinationDir}'. Access forbidden.");
+            }
+        }
+
+//        if ($this->_fileType == IMAGETYPE_JPEG) {
+//            $threshold = (int)Mage::getStoreConfig('catalog/product_image/progressive_threshold');
+//            if ($threshold && $threshold <= (imagesx($this->_imageHandler) * imagesy($this->_imageHandler) / 1000000)) {
+//                imageinterlace($this->_imageHandler, 1);
+//
+//                TODO support progressive JPG? Like this?
+//                $this->_imageHandler->setInterlaceScheme();
+//            }
+//        }
```

Comment:
I don't think anyone really cares about progressive anymore so probably better to delete but not a hard requirement IMHO in this case since it could potentially be tested and used someday..

Reply:
mmm but it seems to handle the whole JPEG processing too, maybe there's no need for anything for jpegs?